File: 1475-maximum-sum-bst-in-binary-tree/maximum-sum-bst-in-binary-tree.java
class BST {
    boolean isBST;
    int sum, min, max;
    public BST(int min, int max, int sum, boolean isBST){
        this.sum = sum;
        this.min = min;
        this.max = max;
        this.isBST = isBST;
    }
}

class Solution {
    private static final int INF = Integer.MAX_VALUE / 2;
    private int maxsum;
    private BST solve(TreeNode root)
    {
        if(root == null){
            return new BST(INF, -INF, 0, true);
        }
        
        var left = solve(root.left);
        var right = solve(root.right);
        
        if(left.isBST && right.isBST && left.max < root.val && right.min > root.val){
            int newMax = Math.max(root.val, right.max);
            int newMin = Math.min(root.val, left.min);
            int totalSum = left.sum + right.sum + root.val;
            maxsum = Math.max(maxsum, totalSum);
            return new BST(newMin, newMax, totalSum, true);
        }
        
        return new BST(left.min, right.max, Math.max(left.sum, right.sum), false);
    }

    public int maxSumBST(TreeNode root)
    {
        maxsum = 0;
        solve(root);
        return maxsum;
    }
}

File: 1475-maximum-sum-bst-in-binary-tree/maximum-sum-bst-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */

class Node {
public:
    int sum, mini, maxi;
    Node(int sum, int mini, int maxi) : sum(sum), mini(mini), maxi(maxi) {}
};

class Solution {
private: 
    int INF, maxSum;
    Node solve(TreeNode* root) 
    {
        if (!root) 
        return Node(0, INF, -INF);
        
        Node left = solve(root->left);
        Node right = solve(root->right);
        
        if(left.maxi < root->val && root->val < right.mini) 
        {
            int sum = left.sum + right.sum + root->val;
            maxSum = max(maxSum, sum);
            return Node(sum, min(left.mini, root->val), max(right.maxi, root->val));
        }
        
        return Node(max(left.sum, right.sum), -INF, INF);
    }
public:
    int maxSumBST(TreeNode* root) 
    {
        INF = INT_MAX/ 2;
        maxSum = -INF;
        solve(root);
        return maxSum < 0 ? 0 : maxSum;
    }
};